Quarkus cannot start with `another operation (install/upgrade/rollback) is in progress`

During startup, the following issue appeared in logs and quarkus cannot start:

```
2024-12-05 15:21:29,434 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step com.iseonline.shared.k8s.deployment.helm.HelmDeployer#startServices threw an exception: java.lang.IllegalStateException: java.lang.IllegalStateException: another operation (install/upgrade/rollback) is in progress
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:540)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:567)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:670)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
        at com.iseonline.shared.k8s.deployment.helm.HelmDeployer.startServices(HelmDeployer.java:87)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.lang.IllegalStateException: another operation (install/upgrade/rollback) is in progress
```

--- a/deployment/pom.xml
+++ b/deployment/pom.xml
@@ -6,15 +6,14 @@
   <parent>
     <groupId>com.ise-online.quarkus</groupId>
     <artifactId>quarkus-dev-apps-k8s-parent</artifactId>
-    <version>0.1.0-SNAPSHOT</version>
+    <version>0.1.2-SNAPSHOT</version>
   </parent>
   <artifactId>quarkus-dev-apps-k8s-deployment</artifactId>
   <name>Quarkus Dev Apps K8s - Deployment</name>
   <dependencies>
     <dependency>
       <groupId>io.fabric8</groupId>
-      <artifactId>openshift-client</artifactId>
-      <version>6.13.3</version>
+      <artifactId>kubernetes-client</artifactId>
     </dependency>
     <dependency>
       <groupId>com.marcnuri.helm-java</groupId>

--- a/deployment/src/main/java/com/iseonline/shared/k8s/deployment/helm/HelmDeployer.java
+++ b/deployment/src/main/java/com/iseonline/shared/k8s/deployment/helm/HelmDeployer.java
@@ -5,6 +5,8 @@ import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.Arrays;
 import java.util.Base64;
 import java.util.Collections;
@@ -16,7 +18,6 @@ import org.jboss.logging.Logger;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
-import com.google.common.base.Strings;
 import com.iseonline.shared.k8s.deployment.K8sDevServicesBuildTimeConfig;
 import com.iseonline.shared.k8s.deployment.ssh.SshDeployer;
 import com.iseonline.shared.k8s.deployment.utils.K8sDevServicesUtils;
@@ -38,6 +39,7 @@ import io.quarkus.deployment.builditem.DevServicesResultBuildItem;
 import io.quarkus.deployment.builditem.DevServicesResultBuildItem.RunningDevService;
 import io.quarkus.deployment.dev.devservices.GlobalDevServicesConfig;
 import io.quarkus.deployment.pkg.builditem.BuildSystemTargetBuildItem;
+import io.quarkus.runtime.util.StringUtil;
 
 @BuildSteps(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class)
 public class HelmDeployer implements Closeable {
@@ -48,6 +50,8 @@ public class HelmDeployer implements Closeable {
     private static final Logger log = Logger.getLogger(HelmDeployer.class);
 
     private static volatile K8sDevServicesBuildTimeConfig config;
+
+    private static volatile KubernetesClient k8sClient;
 
     private static volatile Path kubeConfigPath;
 
@@ -73,6 +77,18 @@ public class HelmDeployer implements Closeable {
         // specify the context to use
         kubeConfigPath = bst.getOutputDirectory().resolve("kubeconfig.yaml");
         saveKubeConfig(config.kubeContext(), kubeConfigPath);
+
+        if (k8sClient == null) {
+            Config k8sConfig = Config.autoConfigure(config.kubeContext());
+            // Nasty hack to avoid having the log be spammed with messages, if multiple
+            // configs are given in env variable, see
+            // https://github.com/fabric8io/kubernetes-client/issues/6240
+            // TODO: Fixed in fabric8, but quarkus must be updated to fabric8 7.0.0
+            k8sConfig.setAutoConfigure(false);
+            k8sClient = new KubernetesClientBuilder()
+                    .withConfig(k8sConfig)
+                    .build();
+        }
 
         Path chartsDir = Path.of(config.chartPath());
         try {
@@ -124,7 +140,7 @@ public class HelmDeployer implements Closeable {
      */
     private void installSecret() {
         String registrySecret = config.registrySecret().orElse(null);
-        if (Strings.isNullOrEmpty(registrySecret)) {
+        if (StringUtil.isNullOrEmpty(registrySecret)) {
             return;
         }
         Config k8sConfig = Config.autoConfigure(config.kubeContext());
@@ -159,7 +175,7 @@ public class HelmDeployer implements Closeable {
 
     private void helmRegistryLogin() {
         String registrySecret = config.registrySecret().orElse(null);
-        if (Strings.isNullOrEmpty(registrySecret)) {
+        if (StringUtil.isNullOrEmpty(registrySecret)) {
             return;
         }
         try {
@@ -219,11 +235,6 @@ public class HelmDeployer implements Closeable {
 
     private void upgradeDeployment(Path chartDir, String releaseName) {
         Helm helm = new Helm(chartDir);
-        long time = System.currentTimeMillis();
-        helm.dependency().update().call();
-        log.debugf("Time for dependency build for helmrelease %s: %d", releaseName,
-                (System.currentTimeMillis() - time));
-        time = System.currentTimeMillis();
         UpgradeCommand upgrade = helm.upgrade();
         for (String profile : profiles) {
             Path profileValuesFile = chartDir.resolve(String.format("values-%s.yaml", profile));
@@ -234,17 +245,49 @@ public class HelmDeployer implements Closeable {
         upgrade.withKubeConfig(kubeConfigPath)
                 .withName(releaseName)
                 .withNamespace(config.namespace())
-                .resetValues()
                 .install()
+                .dependencyUpdate()
                 .createNamespace()
                 // There is currently a bug which prevents the wait, resulting in the error
                 // "beginning wait for resources with timeout of 0s" and "client rate limiter
                 // would exceed context time".
-                // .waitReady()
+                .waitReady()
                 // .debug()
-                .call();
+                .resetValues();
 
-        log.debugf("Time for helm install for helmrelease %s: %d", releaseName, (System.currentTimeMillis() - time));
+        K8sDevServicesUtils.Retry(3, 0, () -> {
+            try {
+                upgrade.call();
+            } catch (Exception e) {
+                if (e.getMessage().contains("another operation (install/upgrade/rollback) is in progress")) {
+                    // As we should be the only ones deploying here, we assume this is an error. Simplest way is to
+                    // delete the corresponding helm secret. As this is only a dev environment, this should not be a problem.
+                    List<Release> releases = Helm.list()
+                            .withNamespace(config.namespace())
+                            .all()
+                            .call();
+                    releases.stream()
+                            .filter(r -> releaseName.equals(r.getName()))
+                            .map(hr -> "sh.helm.release.v1." + releaseName + ".v" + hr.getRevision())
+                            .map(secretName -> k8sClient.secrets()
+                                    .inNamespace(config.namespace())
+                                    .withName(secretName)
+                                    .get())
+                            .filter(secret -> secret != null
+                                    // Magic number: the old release must be at least 60 seconds old, to be sure
+                                    // that there really is no other process in progress.
+                                    // As we don't wait for the release with helm, this should be more than enough
+                                    // time.
+                                    && Instant.parse(secret.getMetadata().getCreationTimestamp())
+                                            .isBefore(Instant.now().minus(Duration.ofSeconds(60))))
+                            .forEach(secret -> k8sClient.secrets()
+                                    .inNamespace(config.namespace())
+                                    .withName(secret.getMetadata().getName())
+                                    .delete());
+                }
+                throw e;
+            }
+        });
     }
 
     private void waitForEverythingReady() {

--- a/deployment/src/main/java/com/iseonline/shared/k8s/deployment/ssh/SshDeployer.java
+++ b/deployment/src/main/java/com/iseonline/shared/k8s/deployment/ssh/SshDeployer.java
@@ -90,7 +90,7 @@ public class SshDeployer implements Closeable {
             // Nasty hack to avoid having the log be spammed with messages, if multiple
             // configs are given in env variable, see
             // https://github.com/fabric8io/kubernetes-client/issues/6240
-            // TODO: If above issue is fixed, this should be removed
+            // TODO: Fixed in fabric8, but quarkus must be updated to fabric8 7.0.0
             k8sConfig.setAutoConfigure(false);
             k8sClient = new KubernetesClientBuilder()
                     .withConfig(k8sConfig)
@@ -229,23 +229,10 @@ public class SshDeployer implements Closeable {
 
             // It can take quite some time before the ssh server is really ready to accept
             // connections, therefore retry some times...
-            int retryCount = 30;
-            for (int i = 0; i < retryCount && session == null; i++) {
-                try {
-                    createK8sTunnel(localSshPort);
-                    createSshSession(config, localSshPort);
-                } catch (Exception e) {
-                    if (i < retryCount - 1) {
-                        log.warnf("Tunnel to k8s cluster failed, retrying %d/%d", (i + 1), retryCount);
-                        try {
-                            Thread.sleep(1000);
-                        } catch (Exception e2) {
-                        }
-                    } else {
-                        throw new RuntimeException(e);
-                    }
-                }
-            }
+            K8sDevServicesUtils.Retry(60, 1000, () -> {
+                createK8sTunnel(localSshPort);
+                createSshSession(config, localSshPort);
+            });
 
             Map<String, String> overrideConfigs = new HashMap<>();
             List<Exception> errors = portsConfg.getPortForwardings().stream()
@@ -317,7 +304,7 @@ public class SshDeployer implements Closeable {
         }, 10, TimeUnit.SECONDS);
     }
 
-    private void createSshSession(K8sDevServicesBuildTimeConfig config, int localSshPort) throws JSchException {
+    private void createSshSession(K8sDevServicesBuildTimeConfig config, int localSshPort) throws RuntimeException {
         K8sDevServicesUtils.createAndWatch(() -> {
             closeSsh();
             log.infof("Connecting ssh on port %d", localSshPort);

--- a/deployment/src/main/java/com/iseonline/shared/k8s/deployment/utils/K8sDevServicesUtils.java
+++ b/deployment/src/main/java/com/iseonline/shared/k8s/deployment/utils/K8sDevServicesUtils.java
@@ -58,4 +58,23 @@ public class K8sDevServicesUtils {
             }
         }, period, period, unit);
     }
+
+    public static void Retry(int maxRetries, int backoff, Runnable retrieable) {
+        for (int i = 0; i < maxRetries; i++) {
+            try {
+                retrieable.run();
+                return;
+            } catch (Exception e) {
+                if (i < maxRetries - 1) {
+                    log.warnf("Retrying %d/%d", (i + 1), maxRetries);
+                    try {
+                        Thread.sleep(backoff);
+                    } catch (Exception e2) {
+                    }
+                } else {
+                    throw new RuntimeException("Backing off retry after " + maxRetries + " attempts", e);
+                }
+            }
+        }
+    }
 }

--- a/deployment/src/test/java/com/iseonline/shared/k8s/deployment/ssh/SshDeployerTest.java
+++ b/deployment/src/test/java/com/iseonline/shared/k8s/deployment/ssh/SshDeployerTest.java
@@ -11,8 +11,6 @@ import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
-import com.iseonline.shared.k8s.deployment.ssh.SshDeployer;
-
 import io.quarkus.test.QuarkusUnitTest;
 
 @Disabled

--- a/pom.xml
+++ b/pom.xml
@@ -11,7 +11,7 @@
   <groupId>com.ise-online.quarkus</groupId>
   <artifactId>quarkus-dev-apps-k8s-parent</artifactId>
   <name>Quarkus Dev Apps K8s - Parent</name>
-  <version>0.1.0-SNAPSHOT</version>
+  <version>0.1.2-SNAPSHOT</version>
   <packaging>pom</packaging>
 
   <description>A Quarkus extension to start dependent services in a Kubernetes cluster during quarkus dev mode</description>

--- a/runtime/pom.xml
+++ b/runtime/pom.xml
@@ -6,15 +6,11 @@
   <parent>
     <groupId>com.ise-online.quarkus</groupId>
     <artifactId>quarkus-dev-apps-k8s-parent</artifactId>
-    <version>0.1.0-SNAPSHOT</version>
+    <version>0.1.2-SNAPSHOT</version>
   </parent>
   <artifactId>quarkus-dev-apps-k8s</artifactId>
   <name>Quarkus Dev Apps K8s Extension</name>
   <dependencies>
-    <dependency>
-      <groupId>io.quarkus</groupId>
-      <artifactId>quarkus-arc</artifactId>
-    </dependency>
     <dependency>
       <groupId>io.quarkus</groupId>
       <artifactId>quarkus-smallrye-reactive-messaging</artifactId>
@@ -24,56 +20,8 @@
       <artifactId>quarkus-jackson</artifactId>
     </dependency>
     <dependency>
-      <groupId>org.graalvm.sdk</groupId>
-      <artifactId>graal-sdk</artifactId>
-      <scope>provided</scope>
-    </dependency>
-    <dependency>
-      <groupId>io.smallrye.reactive</groupId>
-      <artifactId>smallrye-reactive-messaging-otel</artifactId>
-    </dependency>
-    <dependency>
-      <groupId>org.junit.jupiter</groupId>
-      <artifactId>junit-jupiter</artifactId>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>org.assertj</groupId>
-      <artifactId>assertj-core</artifactId>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>com.fasterxml.jackson.module</groupId>
-      <artifactId>jackson-module-jsonSchema</artifactId>
-      <version>2.16.1</version>
-    </dependency>
-    <dependency>
-      <groupId>jakarta.validation</groupId>
-      <artifactId>jakarta.validation-api</artifactId>
-      <version>3.0.2</version>
-    </dependency>
-    <dependency>
-      <groupId>org.projectlombok</groupId>
-      <artifactId>lombok</artifactId>
-      <version>${lombok.version}</version>
-      <scope>provided</scope>
-    </dependency>
-    <dependency>
-      <groupId>com.google.protobuf</groupId>
-      <artifactId>protobuf-java</artifactId>
-    </dependency>
-    <dependency>
-      <groupId>com.google.protobuf</groupId>
-      <artifactId>protobuf-java-util</artifactId>
-    </dependency>
-    <dependency>
       <groupId>io.quarkus</groupId>
       <artifactId>quarkus-micrometer</artifactId>
-    </dependency>
-    <dependency>
-      <groupId>org.reflections</groupId>
-      <artifactId>reflections</artifactId>
-      <version>0.9.12</version>
     </dependency>
   </dependencies>
   <build>

